### ISSUE TITLE
[4.7.2] Issues with actor typing

### ISSUE THREAD

**Issue by Aleksion** [statelyai/xstate#849]
**Description**

`spawn` results in either Actor or Intepreter typing if helper functions are used on the Actor machine

**Expected Result**

I would expect `spawn` to always return the same type of EITHER `Interpreter` or `Actor`.

**Actual Result**

```typescript
spawn(paymentMachine, "payments") // Intepreter
spawn(paymentMachine.withContext({}), "payments") // Actor
```

**Reproduction**
NOTE it only seems to happen if the Actor interface is provided with a context, state tree and event types

- [Reproduction](https://codesandbox.io/embed/inspiring-gates-351r6?fontsize=14&hidenavigation=1&theme=dark)

**Additional context**

**Comment by davidkpiano** [comment]
You can probably cast the type (`... as StateMachine<...>`) as a workaround.

**Comment by Aleksion** [comment]
I've been doing that so far, so I'm not personally blocked by this. So no urgency or expectations from my side.

I would love to jump in and solve it, but I'm a little unsure what the intended typing is:
Is it an Actor or is it an Interpreter? or do the actual underlying implementation ever change based on what's provided to spawn? I would expect spawn to ALWAYS spawn an Actor. If that's the intended behaviour, I can take a stab at this.

I've also noticed that `useActor` and `useService` behaves slightly differently: useService seems to ALWAYS give you an actor/interpreter with a context on it, while useActor seems to have a delay on it (at least in the codeSandbox I shared). I think what's happening is that it only shows a value after a new event has been received, but I might be wrong about this.

**Comment by bradwoods** [comment]
I don't understand why the interpreter type is returned from spawn. The [docs](https://xstate.js.org/api/classes/interpreter.html#spawn) are stating an Actor type should be returned. I did a [code sandbox](https://codesandbox.io/s/actor-type-error-v1b0w) showing the problem in case I missed something.

**Comment by davidkpiano** [comment]
@bradwoods Try with 4.7.3

**Comment by bradwoods** [comment]
Still getting the same type error - https://codesandbox.io/s/actor-type-error-v1b0w

**Comment by JackCA** [comment]
I believe this is happening in[ my example](https://codesandbox.io/s/magical-shape-dw9yz?fontsize=14&hidenavigation=1&theme=dark) as well.

@bradwoods FYI those are 404ing for me

@Aleksion have you modified your original example? It looks like you're getting a different issue now.

Were you getting somethign like I'm seeing in my example?
```
Argument of type '{ children: (ctx: ParentContext) => Interpreter<ChildContext, any, EventObject, any>[]; }' is not assignable to parameter of type 'Assigner<ParentContext, AnyEventObject> | PropertyAssigner<ParentContext, AnyEventObject>'.
  Type '{ children: (ctx: ParentContext) => Interpreter<ChildContext, any, EventObject, any>[]; }' is not assignable to type 'PropertyAssigner<ParentContext, AnyEventObject>'.
    Types of property 'children' are incompatible.
      Type '(ctx: ParentContext) => Interpreter<ChildContext, any, EventObject, any>[]' is not assignable to type 'Actor<ChildContext, ChildEvent>[] | ((context: ParentContext, event: AnyEventObject, meta: AssignMeta<ParentContext, AnyEventObject>) => Actor<ChildContext, ChildEvent>[])'.
        Type '(ctx: ParentContext) => Interpreter<ChildContext, any, EventObject, any>[]' is not assignable to type '(context: ParentContext, event: AnyEventObject, meta: AssignMeta<ParentContext, AnyEventObject>) => Actor<ChildContext, ChildEvent>[]'.
          Type 'Interpreter<ChildContext, any, EventObject, any>[]' is not assignable to type 'Actor<ChildContext, ChildEvent>[]'.
            Type 'Interpreter<ChildContext, any, EventObject, any>' is not assignable to type 'Actor<ChildContext, ChildEvent>'.
              Types of property 'subscribe' are incompatible.
                Type '{ (observer: Observer<State<ChildContext, EventObject, any, any>>): Unsubscribable; (nextListener?: (state: State<ChildContext, EventObject, any, any>) => void, errorListener?: (error: any) => void, completeListener?: () => void): Unsubscribable; }' is not assignable to type '(next?: (value: ChildContext) => void, error?: (error: any) => void, complete?: () => void) => Unsubscribable'.
                  Types of parameters 'observer' and 'next' are incompatible.
                    Type '(value: ChildContext) => void' is missing the following properties from type 'Observer<State<ChildContext, EventObject, any, any>>': next, error, completets(2345)
```
x-state 4.8, typescript 3.8.3

**Comment by bradwoods** [comment]
Sorry about that. Here's the updated link - https://codesandbox.io/s/actor-type-error-m2r91

**Comment by davidkpiano** [comment]
@JackCA Workaround:

```ts
type ChildActor = Actor<ChildContext, ChildEvent> | Interpreter<ChildContext>;
```

**Comment by JackCA** [comment]
@davidkpiano that fixed the above error but later on when I try and reach into an `Actor`, the `state` object is typed as `any`

[Example](https://codesandbox.io/s/x-state-spawn-actor-typing-mcs2k)

**Comment by JackCA** [comment]
Resurfacing this as I've been running into semi-related issues after updating my typescript eslinting setup.

Is there a reason that `StateConfig<TContext,TEvent>`[ isn't used here](https://github.com/davidkpiano/xstate/blob/master/packages/core/src/Actor.ts#L19)? Wouldn't that address this?

**Comment by davidkpiano** [comment]
Yes, an `Actor` can represent anything, not just a spawned machine. The issue in your above example is this: 

```ts
type ChildActor = Actor<ChildContext, ChildEvent> | Interpreter<ChildContext>;
```

You need to either coerce or check that the actor in your `children: ChildActor[]` array is an `Actor` or `Interpreter`.

**Comment by webberwang** [comment]
The `useActor` does not pass generic into `Actor`

https://github.com/davidkpiano/xstate/blob/5fcc6c759365ab870db3dab834abe57ba8c05e50/packages/xstate-react/src/useActor.ts#L5-L7

So context defaults to `any`

https://github.com/davidkpiano/xstate/blob/5fcc6c759365ab870db3dab834abe57ba8c05e50/packages/core/src/Actor.ts#L13-L16

**Comment by davidkpiano** [comment]
The `TContext` is `TEmitted` (i.e., the "emitted" value when subscribing to the actor).

**Comment by webberwang** [comment]
@davidkpiano I mean the argument of `useActor`

```
const actor = {} // <-- always default to TContext = any
const [] = useActor(actor)
```

<img width="723" alt="useActor_d_ts_—_codelab__Workspace_" src="https://user-images.githubusercontent.com/26264925/95665031-71f9c800-0b01-11eb-85d7-9902edb88c20.png">

**Comment by davidkpiano** [comment]
@webberwang can you open up a separate issue for that?

**Comment by j-funk** [comment]
@davidkpiano the `Interpreter` workaround is no longer working - getting the following error when trying to spawn actors in TypeScript. I suspect something has changed from RC 1.0.0 when this was opened to latest.  

Workaround works here:

https://codesandbox.io/s/x-state-spawn-actor-typing-forked-ods1g?file=/src/index.tsx

But does not work here with updated dependencies:

https://codesandbox.io/s/x-state-spawn-actor-typing-forked-91ze9?file=/src/index.tsx

<img width="732" alt="Screen Shot 2021-08-17 at 12 19 36 PM" src="https://user-images.githubusercontent.com/4483540/129763245-54067a01-edf3-4ebe-a28c-98b4e93adc88.png">

**Comment by davidkpiano** [comment]
We need to support inferring an `ActorRef` from an `Interpreter` to make this easier, but the initial example should now be this: https://codesandbox.io/s/keen-grass-qiuz6

The `ActorRef<TEvent, TEmitted>` type should now be used instead of `Interpreter<...>`.
